### PR TITLE
Separate SSH prompting from ENV logic

### PR DESCRIPTION
**Pull request by omus**
Separating some of the SSH callback logic so that `allow_prompt` only enables/disables interactive user prompting. Other changes include:

* Avoid asking for a public key if the private key file does not exist
* Avoid setting "~/.ssh/id_rsa" as the default private key if it does not exist (defaults to "" instead)
* Avoid setting "$private_key.pub" as the default public key if it does not exist (defaults to "" instead)
* Avoid invalidating the public key before the credentials should be "saved"
* Remove private key missing warning. Either the user will be re-prompted or an error will be displayed
* Avoid prompting for a public key as much as possible

Part of #20725.

### CODE REVIEW

File: base/libgit2/callbacks.jl [hunk 1]
```diff
@@ -78,68 +99,40 @@ function authenticate_ssh(libgit2credptr::Ptr{Ptr{Void}}, p::CredentialPayload,
...
-                    keydefpath = defaultkeydefpath
-                else
-                    response = Base.prompt("Private key location for '$prompt_url'",
-                        default=keydefpath)
-                    isnull(response) && return user_abort()
-                    keydefpath = unsafe_get(response)
-                end
+        if !isfile(privatekey)
+            response = Base.prompt("Private key location for '$prompt_url'",
+                default=privatekey)
```
Comment (rofinn) [inline base/libgit2/callbacks.jl]: Might as well fix the spacing to better fit with the existing base julia style (e.g., `default=privatekey` lined up with opening parenthesis).

File: base/libgit2/callbacks.jl [hunk 2]
```diff
@@ -78,68 +99,40 @@ function authenticate_ssh(libgit2credptr::Ptr{Ptr{Void}}, p::CredentialPayload,
...
-                    isnull(response) && return user_abort()
-                    keydefpath = unsafe_get(response)
-                end
-            end
-            keydefpath
+        # For SSH we need a public key location. Avoid asking about the public key as
+        # typically this will just annoy users.
+        if !isfile(publickey) && isfile(privatekey)
+            response = Base.prompt("Public key location for '$prompt_url'",
+                default=publickey)
```
Comment (rofinn) [inline base/libgit2/callbacks.jl]: Same spacing issue.

File: test/libgit2.jl [hunk 3]
```diff
@@ -1825,7 +1825,6 @@ mktempdir() do dir
                 # User provides an empty private key which triggers a re-prompt
                 challenges = [
                     "Private key location for 'git@github.com':" => "\n",
-                    "Public key location for 'git@github.com' [.pub]:" => "\n",
```
Comment (rofinn) [inline test/libgit2.jl]: I understand that this test doesn't make sense now that we're only checking for the public key when the private key is provided, but don't we still need to check the public key prompting when a valid private key is provided?

File: base/libgit2/callbacks.jl [hunk 4]
```diff
@@ -78,68 +99,40 @@ function authenticate_ssh(libgit2credptr::Ptr{Ptr{Void}}, p::CredentialPayload,
...
-                    response = Base.prompt("Passphrase for $privatekey", password=true)
-                    isnull(response) && return user_abort()
-                    passdef = unsafe_get(response)
-                    isempty(passdef) && return user_abort()  # Ambiguous if EOF or newline
-                end
+        if isempty(passphrase) && is_passphrase_required(privatekey)
+            if Sys.iswindows()
+                response = Base.winprompt(
+                    "Your SSH Key requires a password, please enter it now:",
+                    "Passphrase required", privatekey; prompt_username = false)
```
Comment (rofinn) [inline base/libgit2/callbacks.jl]: Spacing.

**Review by rofinn — APPROVED** [review]
Apart from a minor concern around test coverage this seems pretty good. This might also be a good time to fix some style issues with the rest of base, but I don't think it's huge priority.

File: base/libgit2/callbacks.jl (hunk 4 above)
Comment (omus) [reply base/libgit2/callbacks.jl]: Issue here is `prompt_username = false`?

Comment (rofinn) [reply base/libgit2/callbacks.jl]: Yeah, and a few others. Looking at other places in base I'm guessing it should be.

```julia
response = Base.winprompt("Your SSH Key requires a password, ",
                                              "please enter it now: Passphrase required",
                                              privatekey; prompt_username=false)
```

Comment (omus) [reply base/libgit2/callbacks.jl]: One of the more extreme cases of this is in `authenticate_userpass`: 
```julia
response = Base.winprompt("Please enter your credentials for '$prompt_url'", "Credentials required",
    isempty(username) ? p.username : username; prompt_username = true)
```
which would become the following in order to fit within the 92-limit:
```julia
response = Base.winprompt("Please enter your credentials for '$prompt_url'",
                          "Credentials required", 
                          isempty(username) ? p.username : username; 
                          prompt_username=true)
```
I think this is also acceptable in Base (I've been allowed to do this before): 
```julia
response = Base.winprompt(
    "Please enter your credentials for '$prompt_url'", "Credentials required", 
    isempty(username) ? p.username : username; prompt_username=true)
```
I'll note that the last example just slightly exceeds the 92-limit but I think is probably the most readable.

I'll try to strike a balance.

File: test/libgit2.jl (hunk 3 above)
Comment (omus) [reply test/libgit2.jl]: There is one test I can add for public keys at this point. There's one round left of refactoring that will allow me to add a few more tests which currently cause the `credential_loop` to infinite loop.

File: base/libgit2/callbacks.jl (hunk 1 above)
Comment (omus) [reply base/libgit2/callbacks.jl]: I ended up leaving the indentation for the `default` keyword as is as it matches the `winprompt` indentation which is quite long.

**Comment by omus** [comment]
I'll merge once the CI passes.

File: base/libgit2/callbacks.jl (hunk 4 above)
Comment (rofinn) [reply base/libgit2/callbacks.jl]: Thanks.

**Pull request merged**